pkg/rules/rules.go

```diff
@@ -0,0 +1,257 @@
+//     the current subject can perform on which resources in the specified namespace
+//   - error: nil on success, or an error with context if the operation fails
+//
+// The function will return an error if:
+//   - Creating the SelfSubjectRulesReview fails
+//   - The SelfSubjectRulesReview returns an evaluation error
+func RetrieveSelfSubjectRules(
+	ctx context.Context,
+	cli client.Client,
+	ns string,
+) ([]authorizationv1.ResourceRule, error) {
+	rulesReview := authorizationv1.SelfSubjectRulesReview{
+		Spec: authorizationv1.SelfSubjectRulesReviewSpec{
+			Namespace: ns,
+		},
+	}
+
+	err := cli.Create(ctx, &rulesReview)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create SelfSubjectRulesReviews: %w", err)
+	}
+
+	if rulesReview.Status.EvaluationError != "" {
+		return nil, fmt.Errorf("error occurred during rule evaluation: %s", rulesReview.Status.EvaluationError)
+	}
+
+	return rulesReview.Status.ResourceRules, nil
+}
+
+// IsResourceMatchingRule determines if a Kubernetes API resource matches an authorization rule.
+// It checks whether the resource's group and name are covered by the permissions defined in the
+// rule.
+//
+// Parameters:
+//   - resourceGroup: The API group of the resource being checked
+//   - apiRes: The API resource metadata
+//   - rule: The authorization rule to match against
+//
+// Returns:
+//   - bool: true if the resource matches the rule, false otherwise
+func IsResourceMatchingRule(
+	resourceGroup string,
+	apiRes metav1.APIResource,
+	rule authorizationv1.ResourceRule,
+) bool {
+	// Check if the resource group matches any of the rule's API groups
+	for _, ruleGroup := range rule.APIGroups {
+		// Skip if the group doesn't match and isn't a wildcard
+		if resourceGroup != ruleGroup && ruleGroup != ResourceAny {
+			continue
+		}
+
+		// Check if the resource name matches any of the rule's resources
+		for _, ruleResource := range rule.Resources {
+			if apiRes.Name == ruleResource || ruleResource == ResourceAny {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// HasDeletePermission checks if the current subject has permission to delete a specific API
+// resource based on the provided authorization rules.
+//
+// Parameters:
+//   - group: The API group of the resource to check
+//   - apiRes: The API resource metadata
+//   - permissionRules: A slice of authorization rules to check against
+//
+// Returns:
+//   - bool: true if the resource can be deleted by the current subject, false otherwise
+func HasDeletePermission(
+	group string,
+	apiRes metav1.APIResource,
+	permissionRules []authorizationv1.ResourceRule,
+) bool {
+	for _, rule := range permissionRules {
+		// Skip if the rule doesn't grant delete permission
+		if !slices.Contains(rule.Verbs, VerbDelete) && !slices.Contains(rule.Verbs, VerbAny) {
+			continue
+		}
+
+		// Skip if the resource doesn't match this rule
+		if !IsResourceMatchingRule(group, apiRes, rule) {
+			continue
+		}
+
+		return true
```

Comment:
```suggestion
        // Iterate over each permission rule
	for _, rule := range permissionRules {
	        // Check if the rule allows the 'delete' verb or a wildcard for any verb
		if slices.Contains(rule.Verbs, VerbDelete) || slices.Contains(rule.Verbs, VerbAny) {
		        // Check if the current resource matches the rule's group and resource        
			if IsResourceMatchingRule(group, apiRes, rule) {
				return true
			}
		}
	}
	return false

```

Reply:
The reason it is done this way, is to keep the happy path aligned to the left, instead of having nested if 

- https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88
- https://dave.cheney.net/2020/02/23/the-zen-of-go

Reply:
Sounds good, thanks for sharing!